plugins/org.eclipse.elk.alg.layered/src/org/eclipse/elk/alg/layered/intermediate/loops/routing/SplineSelfLoopRouter.java

```diff
@@ -0,0 +1,90 @@
+import org.eclipse.elk.alg.layered.intermediate.loops.SelfLoopPort;
+import org.eclipse.elk.alg.layered.p5edges.splines.NubSpline;
+import org.eclipse.elk.alg.layered.p5edges.splines.SplinesMath;
+import org.eclipse.elk.core.math.KVector;
+import org.eclipse.elk.core.math.KVectorChain;
+import org.eclipse.elk.core.options.PortSide;
+
+/**
+ * Routes self loops with splines.
+ */
+public class SplineSelfLoopRouter extends OrthogonalSelfLoopRouter {
+    
+    /** Dimension of our self loop splines. */
+    private static final int DIM = 3;
+    /** To avoid magic number warning. */
+    private static final double HALF = 0.5;
+    
+    
+    /**
+     * Turns a vector chain of orthogonal bend points into polyline bend points by cutting the corners.
+     */
+    @Override
+    protected KVectorChain modifyBendPoints(final SelfLoopEdge slEdge, final EdgeRoutingDirection routingDirection,
+            final KVectorChain bendPoints) {
+        
+        // For the splines to be routed correctly, we also have to include the source and target positions, so we build
+        // up the new list of bend points bit by bit
+        KVectorChain splineBendPoints = new KVectorChain(relativePortAnchor(slEdge.getSLSource()));
+        addSplineControlPoints(slEdge, routingDirection, bendPoints, splineBendPoints);
+        splineBendPoints.add(relativePortAnchor(slEdge.getSLTarget()));
+        
+        return new NubSpline(true, DIM, splineBendPoints).getBezierCP();
+    }
+    
+    private KVector relativePortAnchor(final SelfLoopPort slPort) {
+        LPort lPort = slPort.getLPort();
+        return new KVector(lPort.getPosition()).add(lPort.getAnchor());
+    }
+    
+    /**
+     * Inserts spline control points between each consecutive pair of bend points as computed by the orthogonal self
+     * loop router. The spline control points are slightly offset away from the node.
+     */
+    private void addSplineControlPoints(final SelfLoopEdge slEdge, final EdgeRoutingDirection routingDirection,
+            final KVectorChain orthoBendPoints, final KVectorChain newBendPoints) {
+        
+        assert orthoBendPoints.size() >= 2;
+        
+        // We want to insert a new bend point between each pair of consecutive bend points in the old list
+        PortSide currPortSide = slEdge.getSLSource().getLPort().getSide();
+        KVector firstBP = orthoBendPoints.get(0);
+        
+        for (int secondBPIndex = 1; secondBPIndex < orthoBendPoints.size(); secondBPIndex++) {
+            // Next pair
+            KVector secondBP = orthoBendPoints.get(secondBPIndex);
+            
+            // The first bend point will go straight into our list before we compute a new bend point
+            newBendPoints.add(firstBP);
+            
+            // Compute a middle bend point and move it away from the node a little
+            KVector midBP = new KVector(firstBP).add(secondBP).scale(HALF);
```

Comment:
Again the question: is this enough for most cases to not overlap labels?

Reply:
The spline will not overlap its own label, but may cross labels of other self loops. Trying to avoid this is hard, so I opted not to attempt it due to time and complexity considerations.